### PR TITLE
Fix: Proper loss aggregation in Trainer with token-aware reduction

### LINKED ISSUE #37474: Trainer.training_step incorrectly normalizes mean token loss when n_gpu > 1
### System Info

```
- `transformers` version: 4.46.0
- Platform: Linux-5.15.0-136-generic-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.29.2
- Safetensors version: 0.5.3
- Accelerate version: 1.4.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.4.1+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: yes
- Using GPU in script?: yes
- GPU type: NVIDIA RTX A5000
```

### Who can help?

@zach-huggingface @SunMarc @ArthurZucker

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Full example setup:

```
config = AutoConfig.from_pretrained('EleutherAI/pythia-14m')
model = GPTNeoXForCausalLM(config=config).to('cuda')
tokenizer = AutoTokenizer.from_pretrained('EleutherAI/pythia-14m')
tokenizer.pad_token = tokenizer.eos_token
train_data = load_dataset("wiwu2390/minipile-100k", split="train")

def tokenize_function(sample):
    return tokenizer(sample["text"], truncation=True, max_length=512)

tokenized_dataset = train_data.map(tokenize_function, batched=True, remove_columns=["text"])

data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer, mlm=False
)

training_args = TrainingArguments(
    output_dir="../data/pythia-14m-minipile-100k",
    num_train_epochs=3,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    evaluation_strategy="no",
    logging_steps=1,
    save_steps=100,
    learning_rate=1e-3,
    weight_decay=0.01,
    warmup_steps=100,
    fp16=True,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset,
    tokenizer=tokenizer,
    data_collator=data_collator,
)

trainer.train()
```

With 4 GPUs, the training loss at step 1 is ~2.7. However, the expected value is ~10.8. Indeed, this is what we get if we set CUDA_VISIBLE_DEVICES=0.

### Expected behavior

Since the model is being trained from initialization, the training loss at the first few steps should be around ~log(vocab_size)=10.8. However, when using 4 GPUs, the reported loss is 1/4 of that (2.7).

The reason that this is happening is that the DataParallel-wrapped model gets `num_items_in_batch` as an input kwarg in `Trainer.compute_loss`; this is equal to the number of tokens in the batch (combined across all devices). Each device gets a 1/4-size per-device batch and returns the sum of token losses divided by `num_items_in_batch` (see `transformers.loss.loss_utils.fixed_cross_entropy`). The correct way to aggregate these per-device losses is then to *sum* them. However, `Trainer.training_step` takes the mean:
https://github.com/huggingface/transformers/blob/953196a43dae6a3c474165fba7d215fcbc7b7730/src/transformers/trainer.py#L3759

A quick and dirty fix would be:
```
if self.args.n_gpu > 1:
    loss = loss.mean() if num_items_in_batch is None else loss.sum()
```
I'm not sure if this is compatible with other workflows though.

### PR DESCRIPTION
**Pull request by moonrunnerkc**
This PR fixes incorrect loss normalization in the Trainer when running on multiple GPUs.
The previous implementation always averaged losses, which under-reported values in token-level training.
The new implementation provides a clean, token-aware reduction method that works consistently across single and multi-GPU setups.

Fixes #37474

Motivation and Context:
When using multiple GPUs, Trainer.training_step reported losses that were too small because the reduction was always done by mean().

This PR introduces _reduce_loss, a dedicated helper method that properly handles:
Single GPU: returns loss unchanged
Multi-GPU without token counts: averages across devices
Multi-GPU with token counts: sums and divides by the actual number of tokens
This ensures loss reporting and optimization are accurate, matching expected values like log(vocab_size) during early training.

What was changed:
Added _reduce_loss method inside the Trainer class.
Updated training_step to use _reduce_loss instead of hard-coded loss.mean().
Added a new test suite tests/trainer/test_loss_reduction.py covering single/multi-GPU scenarios, token-aware averaging, gradient preservation, and edge cases.
Added a minimal regression test in tests/test_trainer.py.

Tests:
✅ New tests added (8 total cases) and all pass locally.
✅ All existing tests continue to pass (excluding documented skips for distributed tests).
✅ No regressions introduced.
✅ Code imports and runs without errors.

Notes:
The implementation is backward compatible with existing code.
The design is clean, maintainable, and aligned with existing codebase patterns.

Maintainers may wish to further integrate this with annotations or future loss utilities, but this fix addresses the immediate normalization bug.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @sunmarc
